packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,202 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+pragma experimental ABIEncoderV2;
+
+import "../../iface/IExchangeV3.sol";
+
+import "../../lib/Claimable.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/ReentrancyGuard.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../lib/EIP712.sol";
+
+
+/// @title Fast withdrawal agent implementation.
+///
+///        Any mediator/cookie jar/... (whoever provides the funds immediately) can call
+///        `executeFastWithdrawals` to provide users immediately with funds onchain.
+///        This allows the security of the funds to be handled by any EOA or smart contract.
+///
+///        Users that want to make use of this functinality have to
+///        authorize this contract as their agent.
+///
+///        TODO: generalize this (here or another agent) with any function call
+///        a user wants to do to be able to simulate completely onchain smart wallet with
+///        funds stored in the Loopring protocol.
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalAgent is ReentrancyGuard
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawal
+    {
+        address exchange;
+        address from;
+        address to;
+        address token;
+        uint    amount;
+        address feeToken;
+        uint    fee;
+        uint    validUntil;
+        bool    onchainFeePayment;
+        bytes   signature;
+    }
+
+    // Replay protection
+    mapping (bytes32 => bool) txHash;
+
+    // EIP712
+    bytes32 constant public FASTWITHDRAWAL_TYPEHASH = keccak256(
+        "FastWithdrawal(address exchange,address from,address to,address token,uint256 amount,address feeToken,uint256 fee,bool onchainFeePayment,uint256 validUntil)"
+    );
+    bytes32 public DOMAIN_SEPARATOR;
+
+    constructor()
+        public
+    {
+        //DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+    }
+
+    function executeFastWithdrawals(FastWithdrawal[] memory fastWithdrawals)
+        public
```

Comment:
Unfortunately this is not possible in solidity because a struct is used (`FastWithdrawal` in this case). I'm not sure what the reason is for that, but solidity has other strange limitations so I can't say I'm surprised. I don't think this is even possible in solidity v0.6, but I could be wrong.

Reply:
Clarification: the thing that is not possible is to go from a `calldata` struct in an external function to a `memory` struct in an internal function (and internal functions can't have `calldata` parameters). This gives an internal compiler error in solidity v0.5 and v0.6. And making a function `external` forces the parameters to be `calldata`.